pkg/controllers/jobset_controller.go

```diff
@@ -143,80 +142,29 @@ func (r *JobSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		Complete(r)
 }
 
-func (r *JobSetReconciler) constructJobsFromTemplate(js *jobset.JobSet, rjob *jobset.ReplicatedJob, ownedJobs *childJobs) ([]*batchv1.Job, error) {
-	var jobs []*batchv1.Job
-
-	// Construct jobs.
-	for i := 0; i < *rjob.Replicas; i++ {
-		// Check if we need to create this job. If not, skip it.
-		jobName := genJobName(js, rjob, i)
-		if create := r.shouldCreateJob(jobName, ownedJobs); !create {
-			continue
-		}
-
-		// Copy labels/annotations to avoid modifying the template itself.
-		labels := make(map[string]string)
-		for k, v := range rjob.Template.Labels {
-			labels[k] = v
-		}
-		annotations := make(map[string]string)
-		for k, v := range rjob.Template.Annotations {
-			annotations[k] = v
-		}
-
-		job := &batchv1.Job{
-			ObjectMeta: metav1.ObjectMeta{
-				Labels:      labels,
-				Annotations: annotations,
-				Name:        genJobName(js, rjob, i),
-				Namespace:   js.Namespace,
-			},
-			Spec: *rjob.Template.Spec.DeepCopy(),
-		}
-
-		// Add job index as a label and annotation.
-		job.Labels[jobset.JobIndexLabel] = strconv.Itoa(i)
-		job.Annotations[jobset.JobIndexLabel] = strconv.Itoa(i)
-
-		// If enableDNSHostnames is set, update job spec to set subdomain as
-		// job name (a headless service with same name as job will be created later).
-		if dnsHostnamesEnabled(rjob) {
-			job.Spec.Template.Spec.Subdomain = job.Name
-		}
-
-		// Set controller owner reference for garbage collection and reconcilation.
-		if err := ctrl.SetControllerReference(js, job, r.Scheme); err != nil {
-			return nil, err
-		}
-
-		jobs = append(jobs, job)
-	}
-	return jobs, nil
-}
-
 // getChildJobs fetches all Jobs owned by the JobSet and returns them
 // categorized by status (active, successful, failed).
-func (r *JobSetReconciler) getChildJobs(ctx context.Context, js *jobset.JobSet, req ctrl.Request) (*childJobs, error) {
+func (r *JobSetReconciler) getChildJobs(ctx context.Context, js *jobset.JobSet) (*childJobs, error) {
 	// Get all active jobs owned by JobSet.
 	var childJobList batchv1.JobList
-	if err := r.List(ctx, &childJobList, client.InNamespace(req.Namespace), client.MatchingFields{jobOwnerKey: req.Name}); err != nil {
+	if err := r.List(ctx, &childJobList, client.InNamespace(js.Namespace), client.MatchingFields{jobOwnerKey: js.Name}); err != nil {
 		return nil, err
 	}
 
-	jobs := childJobs{}
+	ownedJobs := childJobs{}
 	for i, job := range childJobList.Items {
 		_, finishedType := isJobFinished(&job)
 		switch finishedType {
 		case "": // active
-			jobs.active = append(jobs.active, &childJobList.Items[i])
+			ownedJobs.active = append(ownedJobs.active, &childJobList.Items[i])
 		case batchv1.JobFailed:
-			jobs.failed = append(jobs.failed, &childJobList.Items[i])
+			ownedJobs.failed = append(ownedJobs.failed, &childJobList.Items[i])
 		case batchv1.JobComplete:
-			jobs.successful = append(jobs.successful, &childJobList.Items[i])
+			ownedJobs.successful = append(ownedJobs.successful, &childJobList.Items[i])
```

Comment:
active, failed and successful should only include jobs with restart-attempt = restarts

Reply:
those three lists in addition to the jobsToDelete one should be mutually exclusive